### PR TITLE
Feature: Departure distance

### PR DESCRIPTION
**Pull request by RisingLeaf**
## Feature Details
Adds optional departure distances, to force players to stay longer in a system. Can be used to make pirate attacks more dangerous.

## UI Screenshots
https://user-images.githubusercontent.com/85687254/200139602-204e9c66-932c-4572-b5e0-e9d8a9e1eb53.mp4

## Usage Examples
For both, hyper and jump drive being the same:
```
system name
    departure 200
```
For specific jump/hyperdrive distances:
```
system name
    departure
        link 200
        jump 100
```
## Testing Done
see screenshot

## Performance Impact
N/A

## Optional things to do:

- [ ] Use velocity in closest departure point calculation
- [x] Add departure distance for jump drives

### CODE REVIEW

File: source/Ship.cpp
Comment (petervdmeer) [inline source/Ship.cpp:2857]: Some debug-code was left behind here.

File: data/map.txt
Comment (Hurleveur) [inline data/map.txt:22494]: idk why you change the habitable range here
also I guess this is test data so prob best to remove it from the PR since it isnt a draft

File: source/Ship.cpp
Comment (Hurleveur) [inline source/Ship.cpp:2860]: I think you forgot to change this since your tests?
we definitely should have a warning message for the player saying "moving to the closest jumping point" or smth of the like

File: data/map.txt
Comment (petervdmeer) [inline data/map.txt]: Do we want to have this in-game for Rutilicus? Or is this a testing-artefact that got left behind?

File: source/System.h
Comment (Hurleveur) [inline source/System.h:106]: It would probably make sense for this to consider JD and Hyper differently like it happens with the arrival distance? Not sure what is going on with that exactly though

File: data/map.txt
Comment (warp-core) [inline data/map.txt:22494]: Is this just for testing? Provide a savefile with this change made to the system in the `changes`.

File: source/AI.cpp
Comment (petervdmeer) [inline source/AI.cpp:1983]: Does this also work if the Ship is at position (0,0)? And would it make sense to add a (named) constant instead of multiplying with 1.1 to ensure we get outside the range? Very large departure distances would give quite some overshoot with a factor of 1.1.

**Review by Hurleveur — CHANGES_REQUESTED** [review]
I love the concept but this looks a bit unfinished
Maybe some engines should not be affected by this currently? Or, in the future? I can understand it makes sense not to jump too close to stars and all but this could totally change how one travels by Jump Drive over long distances by making it take ages... Maybe they shpuld be considered differently... Not sure about balance things as its not my area

File: source/Ship.cpp
Comment (warp-core) [inline source/Ship.cpp:50]: Debug code?

File: data/map.txt
Comment (RisingLeaf) [reply data/map.txt:22494]: whoops my testing is in

File: source/Ship.cpp
Comment (petervdmeer) [inline source/Ship.cpp:50]: Testing artifact.

File: source/Ship.cpp
Comment (warp-core) [inline source/Ship.cpp:2858]: Use `LengthSquared()` and square the departure distance.

File: source/System.h
Comment (RisingLeaf) [reply source/System.h:106]: I can do so

File: source/AI.cpp
Comment (warp-core) [inline source/AI.cpp:1981]: Again, square.

File: source/AI.cpp
Comment (warp-core) [inline source/AI.cpp:1983]: square

**Review by petervdmeer — COMMENTED** [review]
The code looks good.

Some testing artifacts were left in the code, and I had one minor remark on the actual distance to get out of the departure zone.

File: source/AI.cpp
Comment (RisingLeaf) [reply source/AI.cpp:1983]: yeah, this was for testing, I can remove it now

**Comment by RisingLeaf** [comment]
Removed testing and implemented use of squares

File: source/System.h
Comment (petervdmeer) [reply source/System.h:106]: We actually have [an RFC in progress](https://github.com/EndlessSkyCommunity/endless-sky/pull/51/files) where syntax for jump and hyperdrives were discussed. But don't worry about RFCs in progress; they are not official directions until they are officially merged.

Comment (RisingLeaf) [reply source/System.h:106]: Hmmm... I can try to implement some of the things there, I dont see the use of a max though

Comment (petervdmeer) [reply source/System.h:106]: Please do keep things small and simple and just focus on basic departure distance in this PR. Differentiating between hyperdrives and jumpdrives can also be done in later PRs. (And just refer to the RFC for questions on how that can be done later.)

Comment (RisingLeaf) [reply source/System.h]: sorry, too late

**Review by UnorderedSigh — COMMENTED** [review]
I don't see a visual indication of the departure distance, so I think it'll confuse the player. If this is to be used widely, it should be added to the tutorial. However, even without a visual indication, I see niche uses in missions: "due to the gravitation pull of the black hole, you must travel a long distance away before jumping."

Also, in the early game, orbit camping is deadly enough already. This can be tuned on a per-system basis, so that's good.

I wouldn't consider this work 100% complete until there is a visual indication and tutorial update. However, it would be reasonable to separate a content PR from a code PR, ~~so I'll let others decide whether this should be merged as is.~~

Edit: My comment about others deciding was explaining why I'm not approving or requesting changes. I wasn't claiming dictatorial control over anything. Sorry for the misunderstanding.

**Comment by EjoThims** [comment]
Visual indicator in every system would be a huge mess.

Specific hazards can be used to create indicators for areas in which it specifically increased.

I like the idea of having the ability to specify between the drives as well, even though I can't think of anywhere it would be appropriate in current vanilla content. I love supporting future work and plug-in additions as part of code enhancements :)

**Comment by Hecter94** [comment]
> I don't see a visual indication of the departure distance, so I think it'll confuse the player. 

I don't think there's any need to have an always-on visual indicator.
We don't have a visual indicator of arrival distance and it doesn't seem to confuse anyone. It's also not a terribly difficult concept to understand. 

At the most, I could see a visual indication that briefly pulses on if the player attempts to jump while still within the departure distance.

> If this is to be used widely, it should be added to the tutorial. 

Yes, it should be.

**Comment by UnorderedSigh** [comment]
> We don't have a visual indicator of arrival distance and it doesn't seem to confuse anyone. It's also not a terribly difficult concept to understand.

If you cannot depart a system before a certain distance, then you don't know how far you need to fly out when fleeing in combat. Not knowing where to go will make survival much more difficult when you're outgunned and need to retreat.

**Comment by UnorderedSigh** [comment]
> There's also no reason to overhaul the tutorial if autopilot accounts for it.

The autopilot doesn't know how to dodge shots while flying to a place where you can safely jump. The player has to do that, and can't if they have no way of knowing where they need to be to jump.

**Comment by Hurleveur** [comment]
It could just be shown on the radar?

**Comment by UnorderedSigh** [comment]
I think it would be best to add just the departure distance mechanics in this PR, and work on the content and UI changes in other PRs. Keep game mechanics, content, and UI separate. Departure distance does no harm since it isn't used yet.

File: source/AI.cpp
Comment (petervdmeer) [inline source/AI.cpp:1984]: If we compare to the squaredDeparture, then I guess we should also use `ship.Position().LengthSquared()`, otherwise we move quite a distance beyond the intended departure point..

File: source/AI.cpp
Comment (petervdmeer) [reply source/AI.cpp]: I'm afraid we still have a division by zero if the Ship is at position (0,0), because I would expect both Length() and LengthSquared() to return 0 in such cases.
And even if we would not have a division by zero, the calculation of the closestDeparturePoint would still result in position(0,0), because multiplying zero with anything doesn't give any new point.

File: source/AI.cpp
Comment (petervdmeer) [inline source/AI.cpp:1981]: Let's overshoot a little bit by adding some (named) constant; we don't want AI ships to "hang at the jump border" because they move to a point that just doesn't pass the check on line 1984 by some tiny value.

File: source/System.cpp
Comment (petervdmeer) [inline source/System.cpp:356]: Does this work for the "specific distances" syntax, because there I expect hasValue to be false if the values are given specifically for link and for jump.

File: source/AI.cpp
Comment (RisingLeaf) [reply source/AI.cpp]: done

File: source/AI.cpp
Comment (RisingLeaf) [reply source/AI.cpp:1983]: done

File: source/System.cpp
Comment (RisingLeaf) [reply source/System.cpp]: fixed

File: source/AI.cpp
Comment (RisingLeaf) [reply source/AI.cpp]: fixed

File: source/System.h
Comment (quyykk) [inline source/System.h:106]: Add a comment please :)

File: source/Ship.cpp
Comment (quyykk) [inline source/Ship.cpp:2869]: I think you can leave the condition on one line? idk

File: source/AI.cpp
Comment (quyykk) [inline source/AI.cpp:1989]: The 0.1 should definitely have a comment and maybe also be a named constant.

File: source/System.h
Comment (RisingLeaf) [reply source/System.h:106]: will do the comments

File: source/AI.cpp
Comment (RisingLeaf) [reply source/AI.cpp]: done

File: source/AI.cpp
Comment (petervdmeer) [inline source/AI.cpp:1991]: The division by zero is now fixed, but if ship.Position() is (0,0), then the multiplication will still not result in another position than 0,0.

Comment (RisingLeaf) [reply source/AI.cpp:1991]: you are right...

Comment (RisingLeaf) [reply source/AI.cpp]: fixed

File: source/AI.cpp
Comment (petervdmeer) [inline source/AI.cpp:1990]: `ship.Position()` can be translated to bool directly, no need to perform an expensive Pythagoras operation to get the `Length()` here.
```suggestion
		if(ship.Position())
```

File: source/AI.cpp
Comment (petervdmeer) [inline source/AI.cpp:1994]: No need to calculate a point here, it is possible to immediately generate one. (But it doesn't really matter much, this code gets rarely executed.)
```suggestion
			closestDeparturePoint = Point(1., squaredDeparture);
```

**Review by petervdmeer — APPROVED** [review]
All my review comments appear to be handled. (Except maybe a very small minor comment, but that one can be ignored.)

**Comment by Zitchas** [comment]
In the EV series, where hyperspace was a much more manual experience, the GUI had a "current hyperspace destination" box that was fairly prominent. This text lit up when the player was sufficiently far out to jump. 

I agree, though, that this PR should just focus on the mechanic itself, with GUI and indicators being a separate thing. The above is just an idea for future consideration.

Thanks for making this!

File: source/AI.cpp
Comment (RisingLeaf) [reply source/AI.cpp:292]: done

File: source/Ship.cpp
Comment (RisingLeaf) [reply source/Ship.cpp:2866]: I did something, I want to hold on the variable squared departure though, because I use it three times.

Comment (quyykk) [reply source/Ship.cpp:2890]: Works for me! Oh you changed the code above this one, but not in this file.

File: source/AI.cpp
Comment (quyykk) [inline source/AI.cpp:1988]: You can move the add into the definition above now.

Comment (RisingLeaf) [reply source/AI.cpp:1988]: ok

File: source/Ship.cpp
Comment (RisingLeaf) [reply source/Ship.cpp]: done

File: source/AI.cpp
Comment (RisingLeaf) [reply source/AI.cpp]: done

File: source/Ship.cpp
Comment (quyykk) [inline source/Ship.cpp:2894]: ~~last change I swear~~ can you add a comment here, maybe something like "If the system has a departure distance, the ship can only enter hyperspace if it is beyond that distance from the system center."

Comment (RisingLeaf) [reply source/Ship.cpp:2896]: <b>if it is the last one</b>
done

File: source/AI.cpp
Comment (quyykk) [inline source/AI.cpp:1996]:
```suggestion
		MoveTo(ship, command, closestDeparturePoint, Point(), 0., 0.);
```

~~actually the last one now~~

Comment (RisingLeaf) [reply source/AI.cpp]: ...

**Review by quyykk — APPROVED** [review]

File: source/System.cpp
Comment (quyykk) [inline source/System.cpp:356]:
```suggestion
		else if(key == "departure")
```

**Pull request merged**
